libraries/botframework-schema/src/ActivityEx.ts

```diff
@@ -0,0 +1,342 @@
+        }
+
+        if (reference.locale !== undefined) {
+            source.locale = reference.locale;
+        }
+
+        if(isComming) {
+            if (reference.user !== undefined) {
+                source.from = reference.user;
+            }
+
+            if (reference.bot !== undefined) {
+                source.recipient = reference.bot;
+            }
+
+            if(reference.activityId !== undefined) {
+                source.id = reference.activityId;
+            }
+        } else {
+            if(reference.bot !== undefined) {
+                source.from = reference.bot;
+            }
+            if(reference.user !== undefined) {
+                source.recipient = reference.user;
+            }
+            if(reference.activityId !== undefined) {
+                source.replyToId = reference.activityId;
+            }
+        }
+
+        return source;
+    }
+
+    export function IsFromStreamingConnection(source: Partial<Activity>) : boolean {
+        let isHttp = source.serviceUrl.toLowerCase().startsWith("http");
+
+        return isHttp? !isHttp : false;
+    }
+
+    export function getContinuationActivity(source: Partial<ConversationReference>): Partial<Activity> {
+        if (source === undefined) {
+            throw new Error('source needs to be defined');
+        }
+
+        return {
+            type: ActivityTypes.Event,
+            name: 'ContinueConversation',
+            channelId: source.channelId,
+            serviceUrl: source.serviceUrl,
+            conversation: source.conversation,
+            recipient: source.bot,
+            from: source.user,
+            relatesTo: source as ConversationReference
+        };
+    }
+
+    export function isStartActivity(activity: Activity): boolean {
+        switch (activity.channelId) {
+            case Channels.Skype: {
+                if (activity.type === ActivityTypes.ContactRelationUpdate && activity.action === 'add') {
+                    return true;
+                }
+
+                return false;
+            }
+            case Channels.Directline:
+            case Channels.Emulator:
+            case Channels.Webchat:
+            case Channels.Msteams:
+            case Channels.DirectlineSpeech:
+            case Channels.Test: {
+                if (activity.type === ActivityTypes.ConversationUpdate) {
+                    // When bot is added to the conversation (triggers start only once per conversation)
+                    if (activity.membersAdded !== undefined &&
+                        activity.membersAdded.some((m: ChannelAccount): boolean => m.id === activity.recipient.id)) {
+                        return true;
+                    }
+                }
+
+                return false;
+            }
+            default:
+                return false;
+        }
+    }
+
+    export function isActivity(source:Partial<Activity>, activityType: string): boolean {
+        let type = source.type;

```

Comment:
The variable could be a `const`, as its entire value doesn't change.

Reply:
Feedback applied. Thanks!